grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java

```diff
@@ -0,0 +1,215 @@
+ */
+final class UnaryServerCall<I, O> extends AbstractServerCall<I, O> {
+
+    private static final Logger logger = LoggerFactory.getLogger(UnaryServerCall.class);
+
+    private final HttpRequest req;
+    private final CompletableFuture<HttpResponse> resFuture;
+    private final ServiceRequestContext ctx;
+    private final UnaryMessageDeframer requestDeframer;
+
+    // Only set once.
+    @Nullable
+    private O responseMessage;
+    private boolean deframingStarted;
+
+    UnaryServerCall(HttpRequest req, MethodDescriptor<I, O> method, String simpleMethodName,
+                    CompressorRegistry compressorRegistry, DecompressorRegistry decompressorRegistry,
+                    HttpResponse res, CompletableFuture<HttpResponse> resFuture,
+                    int maxRequestMessageLength, int maxResponseMessageLength,
+                    ServiceRequestContext ctx, SerializationFormat serializationFormat,
+                    @Nullable GrpcJsonMarshaller jsonMarshaller, boolean unsafeWrapRequestBuffers,
+                    boolean useBlockingTaskExecutor, ResponseHeaders defaultHeaders,
+                    @Nullable GrpcStatusFunction statusFunction, boolean autoCompress) {
+        super(req, method, simpleMethodName, compressorRegistry, decompressorRegistry, res,
+              maxResponseMessageLength, ctx, serializationFormat, jsonMarshaller, unsafeWrapRequestBuffers,
+              useBlockingTaskExecutor, defaultHeaders, statusFunction, autoCompress);
+        requireNonNull(req, "req");
+        this.ctx = requireNonNull(ctx, "ctx");
+        final boolean grpcWebText = GrpcSerializationFormats.isGrpcWebText(serializationFormat);
+        requireNonNull(decompressorRegistry, "decompressorRegistry");
+
+        final RequestHeaders clientHeaders = req.headers();
+        requestDeframer = new UnaryMessageDeframer(ctx.alloc(), maxRequestMessageLength, grpcWebText)
+                .decompressor(clientDecompressor(clientHeaders, decompressorRegistry));
+        this.req = req;
+        this.resFuture = requireNonNull(resFuture, "resFuture");
+    }
+
+    @Override
+    public void request(int numMessages) {
+        if (ctx.eventLoop().inEventLoop()) {
+            request0();
+        } else {
+            ctx.eventLoop().execute(this::request0);
+        }
+    }
+
+    private void request0() {
+        if (listener() == null) {
+            return;
+        }
+        startDeframing();
+    }
+
+    @Override
+    void startDeframing() {
+        if (deframingStarted) {
+            return;
+        }
+        deframingStarted = true;
+        req.collect(ctx.eventLoop(), SubscriptionOption.WITH_POOLED_OBJECTS).handle((objects, cause) -> {
+            if (cause != null) {
+                onError(cause);
+                return null;
+            }
+
+            try {
+                onRequestMessage(deframe(objects), true);
+            } catch (Exception ex) {
+                onError(ex);
```

Comment:
We don't need to release the data here.
Because an exception is raised when a deframer detects malformed data
which is released by the try-with-resource block. 
Anyway, will leave an explanation for that.

Reply:
https://github.com/line/armeria/blob/0c5692de73b005cb8a124278ab37bf56b7bd3b99/grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryMessageDeframer.java#L64